### PR TITLE
Refactor `drop columns` to fix issues

### PR DESCRIPTION
**Pull request by IanManske**
# Description
This PR refactors `drop columns` and fixes issues #10902 and #6846. Tables with "holes" are now handled consistently, although still somewhat awkwardly. That is, the columns in the first row are used to determine which columns to drop, meaning that the columns displayed all the way to the right by `table` may not be the columns actually being dropped. For example, `[{a: 1}, {b: 2}] | drop column` will drop column `a` instead of `b`. Before, this would give a list of empty records.

# User-Facing Changes
`drop columns` can now take records as input.

### CODE REVIEW

**Review by sholderbach — COMMENTED** [review]
Thanks for the refactor!

I am not sure what the expectations of folks are for jagged tables, so would be interested what folks expect. (In most of those cases `reject` is probably preferrable for practical purposes)

File: crates/nu-command/src/filters/drop/column.rs [hunk 1]
```diff
@@ -41,148 +46,148 @@ impl Command for DropColumn {
...
+    val: &mut Value,
+    head: Span,
+    drop_cols: &HashSet<String>,
+) -> Result<(), ShellError> {
+    if let Value::Record { val, .. } = val {
+        // TODO: Needs `Record::retain` to be performant,
+        // since this is currently O(n^2)
+        // where n is the number of columns being dropped.
+        // (Assuming dropped columns are at the end of the record.)
+        val.retain(|col, _| !drop_cols.contains(col));
```
[the remark this answers is not in the document]
Comment (IanManske) [reply crates/nu-command/src/filters/drop/column.rs:169]: No, I agree. The comment is probably unclear. In this case, I meant `n` is the number of columns going to be dropped and not the total number of columns.

File: crates/nu-command/src/filters/drop/column.rs [hunk 2]
```diff
@@ -41,148 +46,148 @@ impl Command for DropColumn {
...
-fn get_input_cols(input: Vec<Value>) -> Vec<String> {
-    let rec = input.first();
-    match rec {
-        Some(Value::Record { val, .. }) => val.cols.to_vec(),
-        _ => vec!["".to_string()],
+fn drop_cols_set(val: &mut Value, head: Span, drop: usize) -> Result<HashSet<String>, ShellError> {
+    if let Value::Record { val: record, .. } = val {
+        let len = record.len().saturating_sub(drop);
+        record.vals.truncate(len);
+        Ok(record.cols.drain(len..).collect())
```
[the remark this answers is not in the document]
Comment (IanManske) [reply crates/nu-command/src/filters/drop/column.rs:153]: Yep, maybe also `Record::truncate` for line 130 (but drain should work as well).

**Review by sholderbach — APPROVED** [review]
Rerunning the failed job.

I think I don't want to block this based on the behavior on tables with holes for now.
Great that this PR unlocks records and moves us to the sane API.

We can close #6846 with this and keep #10902 open I think.

I think I would like to bias us to stream by default so we may have to live with the weird behavior. We could consider adding a flag that collects the table so you can get sane truncation for tables with sparse columns.

**Pull request merged**